src/Css/Style.php

```diff
@@ -1254,6 +1283,20 @@ public function inherit(?Style $parent = null): void
                     $this->_props_used[$prop] = null;
                 }
             }
+
+            // Iterate over the parent properties and inherit all the custom
+            // properties that are not defined on the own element.
+            foreach ($parent->_props as $parent_prop => $parent_val) {
```

Comment:
It think it makes sense to combine this loop with the one above:

* Iterate over all parent props.
* The shorthand check can go, since shorthand values are not stored in `_props`.
* Skip, if the property is set on the current element, does not inherit by default, or is not a custom property.
* `Style::$_inherited` could be turned into a map (`[$prop => true]`) to make the inherit-by-default check more efficient.

Reply:
I have done the preparatory work in #3322. With that, combining the loops should be trivial.

Reply:
> The shorthand check can go, since shorthand values are not stored in _props.

I think we're going to have to start storing the shorthand properties so that we can re-calculate when the value of a custom property changes. I don't see an obvious way to make this work otherwise, because you don't know until after the variable has resolved which specific property it applies to.

Reply:
Reference the following sample:

```
<style>
    :root {
        --color: orange;
    }
    div {
        font-size: 150px;
        border: 2px solid var(--color);
    }
</style>
<div>dompdf</div>
```

Since the custom property isn't defined until inheritance is applied the border property parses as invalid.

In order to work around the issue I had to add two steps:

1. pre-processing variables before calling the shorthand method
2. when a custom property is set re-compute all standard properties in the style

The latter, while non-optimal from a performance perspective, is necessary because a specific property is overridden when re-calculating any variable-dependent properties. Such as in the following:

```
<style>
    :root {
        --color: orange;
    }
    div {
        font-size: 150px;
        border: 2px solid var(--color);
        border-left-color: black;
    }
</style>
<div>dompdf</div>
```

I'm happy to push the adjustments I tried locally that appear to resolve current issues.

Reply:
Combined the inheritance logic. Leaving open for Brian's last comment.